pkg/sdn/plugin/node.go

```diff
@@ -125,18 +149,140 @@ func (node *OsdnNode) Start() error {
 		}
 		for _, p := range pods {
 			containerID := getPodContainerID(&p)
-			err = node.UpdatePod(p.Namespace, p.Name, kubeletTypes.DockerID(containerID))
+			err = node.UpdatePod(p.Namespace, p.Name, kubeletTypes.ContainerID{ID: containerID})
 			if err != nil {
 				log.Warningf("Could not update pod %q (%s): %s", p.Name, containerID, err)
 			}
 		}
 	}
 
+	if err := node.cniServer.Start(); err != nil {
+		return err
+	}
+	go utilwait.Forever(node.watchCni, 0)
+
 	node.markPodNetworkReady()
 
 	return nil
 }
 
+type podWorker struct {
+	node     *OsdnNode
+	requests chan *podRequest
+}
+
+func newPodWorker(node *OsdnNode) *podWorker {
+	return &podWorker{
+		node: node,
+		// Allow queue depth of 10 before blocking additions, if there
+		// are more than 10 outstanding requests something is wrong
+		requests: make(chan *podRequest, 10),
+	}
+}
+
+func (w *podWorker) addRequest(request *podRequest) {
+	w.requests <- request
+	if len(w.requests) == 1 {
+		// Start proessing when the first request is added
+		go w.process()
+	}
+}
+
+func (w *podWorker) process() {
+	for request := range w.requests {
+		log.V(5).Infof("Handling pod network request %#v", request)
+
+		var err error
+		result := &podResult{code: http.StatusOK}
+		switch request.command {
+		case CNI_ADD:
+			var ipamResult *cnitypes.Result
+			ipamResult, err = w.node.setupPod(request)
+			if err == nil {
+				result.responseJson, err = json.Marshal(ipamResult)
+			}
+		case CNI_DEL:
+			err = w.node.teardownPod(request)
+		case CNI_UPDATE:
+			err = w.node.updatePod(request)
+		}
+
+		if err != nil {
+			result.code = http.StatusBadRequest
+			result.err = fmt.Sprintf("%v", err)
+		}
+		log.V(5).Infof("Returning pod network result %#v", result)
+		request.result <- result
+	}
+}
+
+func getPodKey(request *podRequest) string {
+	return fmt.Sprintf("%s/%s", request.podNamespace, request.podName)
+}
+
+func (node *OsdnNode) watchCni() {
```

Comment:
methods like this are almost always called "run" or "handle" some variant.  I would suggest you call this `handleCNIRequests` and you should also provide thorough Godoc describing the loop you are adding here.